"error occurred during inference" MethodError on 0.6RC1

Apologies I don't have a reduced code snippet, but this is happening in a pretty big chunk of code and it requires restarting Julia every time to narrow down the actual source of the error. Instead, I thought maybe just maybe someone knowledgeable enough can guess what's happening from the stack trace below, or offer guesses as to what to look for in narrowing it down. 

Basically, once I execute the offending code, the below error/warning is printed, although the program continues seemingly without error. Its reproducible every time. This is on 0.6rc1. 

```
WARNING: An error occurred during inference. Type inference is now partially disabled.
Base.MethodError(f=typeof(Base.string)(), args=(Expr(:||, Expr(:call, :!, Expr(:call, :isa, :t, :Type)::Any)::Any, Expr(:<:, :t, :r)::Any)::Any,), world=0x0000000000000abf)
rec_backtrace at /home/marius/src/julia/src/stackwalk.c:84
record_backtrace at /home/marius/src/julia/src/task.c:245
jl_throw at /home/marius/src/julia/src/task.c:564
jl_method_error_bare at /home/marius/src/julia/src/gf.c:1476
jl_method_error at /home/marius/src/julia/src/gf.c:1494
jl_lookup_generic_ at /home/marius/src/julia/src/gf.c:1907 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1927
limit_type_depth at ./inference.jl:666
unknown function (ip: 0x7fe0b3770f92)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1318
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2760
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
copy! at ./abstractarray.jl:565
abstract_eval_call at ./inference.jl:1868
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_interpret at ./inference.jl:2043
unknown function (ip: 0x7fe0b3742236)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2707
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2760
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_interpret at ./inference.jl:2043
unknown function (ip: 0x7fe0b3742236)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2707
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2760
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2760
typeinf_loop at ./inference.jl:2623
typeinf_frame at ./inference.jl:2476
typeinf_code at ./inference.jl:2546
unknown function (ip: 0x7fe0b375d59d)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_ext at ./inference.jl:2585
unknown function (ip: 0x7fe0b373ca02)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
jl_apply at /home/marius/src/julia/src/julia.h:1422 [inlined]
jl_apply_with_saved_exception_state at /home/marius/src/julia/src/rtutils.c:257
jl_type_infer at /home/marius/src/julia/src/gf.c:269
jl_compile_for_dispatch at /home/marius/src/julia/src/gf.c:1665
jl_compile_method_internal at /home/marius/src/julia/src/julia_internal.h:305 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:352 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
do_call at /home/marius/src/julia/src/interpreter.c:75
eval at /home/marius/src/julia/src/interpreter.c:242
jl_interpret_toplevel_expr at /home/marius/src/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/marius/src/julia/src/toplevel.c:575
jl_parse_eval_all at /home/marius/src/julia/src/ast.c:873
include_string at ./loading.jl:498
include_string at /home/marius/.julia/v0.6/CodeTools/src/eval.jl:30
unknown function (ip: 0x7fe04d48f9b0)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
do_call at /home/marius/src/julia/src/interpreter.c:75
eval at /home/marius/src/julia/src/interpreter.c:242
jl_interpret_toplevel_expr at /home/marius/src/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/marius/src/julia/src/toplevel.c:575
jl_toplevel_eval_in at /home/marius/src/julia/src/builtins.c:496
include_string at /home/marius/.julia/v0.6/CodeTools/src/eval.jl:34
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
#49 at /home/marius/.julia/v0.6/Atom/src/eval.jl:50
withpath at /home/marius/.julia/v0.6/CodeTools/src/utils.jl:30
unknown function (ip: 0x7fe04d48f6a2)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
withpath at /home/marius/.julia/v0.6/Atom/src/eval.jl:38
unknown function (ip: 0x7fe04d48f386)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930WARNING: An error occurred during inference. Type inference is now partially disabled.
Base.MethodError(f=typeof(Base.string)(), args=(Expr(:||, Expr(:call, :!, Expr(:call, :isa, :t, :Type)::Any)::Any, Expr(:<:, :t, :r)::Any)::Any,), world=0x0000000000000abf)
rec_backtrace at /home/marius/src/julia/src/stackwalk.c:84
record_backtrace at /home/marius/src/julia/src/task.c:245
jl_throw at /home/marius/src/julia/src/task.c:564
jl_method_error_bare at /home/marius/src/julia/src/gf.c:1476
jl_method_error at /home/marius/src/julia/src/gf.c:1494
jl_lookup_generic_ at /home/marius/src/julia/src/gf.c:1907 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1927
limit_type_depth at ./inference.jl:666
unknown function (ip: 0x7fe0b3770f92)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1318
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2760
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
copy! at ./abstractarray.jl:565
abstract_eval_call at ./inference.jl:1868
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_interpret at ./inference.jl:2043
unknown function (ip: 0x7fe0b3742236)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2707
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2760
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_interpret at ./inference.jl:2043
unknown function (ip: 0x7fe0b3742236)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2707
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2760
typeinf_loop at ./inference.jl:2606
typeinf_frame at ./inference.jl:2476
typeinf_edge at ./inference.jl:2499
unknown function (ip: 0x7fe0b374ce5a)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call_gf_by_type at ./inference.jl:1386
unknown function (ip: 0x7fe0b374adc6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_call at ./inference.jl:1864
unknown function (ip: 0x7fe0b3746cde)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
abstract_eval_call at ./inference.jl:1894
abstract_eval at ./inference.jl:1917
unknown function (ip: 0x7fe0b3741be6)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_frame at ./inference.jl:2760
typeinf_loop at ./inference.jl:2623
typeinf_frame at ./inference.jl:2476
typeinf_code at ./inference.jl:2546
unknown function (ip: 0x7fe0b375d59d)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
typeinf_ext at ./inference.jl:2585
unknown function (ip: 0x7fe0b373ca02)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
jl_apply at /home/marius/src/julia/src/julia.h:1422 [inlined]
jl_apply_with_saved_exception_state at /home/marius/src/julia/src/rtutils.c:257
jl_type_infer at /home/marius/src/julia/src/gf.c:269
jl_compile_for_dispatch at /home/marius/src/julia/src/gf.c:1665
jl_compile_method_internal at /home/marius/src/julia/src/julia_internal.h:305 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:352 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
do_call at /home/marius/src/julia/src/interpreter.c:75
eval at /home/marius/src/julia/src/interpreter.c:242
jl_interpret_toplevel_expr at /home/marius/src/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/marius/src/julia/src/toplevel.c:575
jl_parse_eval_all at /home/marius/src/julia/src/ast.c:873
include_string at ./loading.jl:498
include_string at /home/marius/.julia/v0.6/CodeTools/src/eval.jl:30
unknown function (ip: 0x7fe04d48f9b0)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
do_call at /home/marius/src/julia/src/interpreter.c:75
eval at /home/marius/src/julia/src/interpreter.c:242
jl_interpret_toplevel_expr at /home/marius/src/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/marius/src/julia/src/toplevel.c:575
jl_toplevel_eval_in at /home/marius/src/julia/src/builtins.c:496
include_string at /home/marius/.julia/v0.6/CodeTools/src/eval.jl:34
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
#49 at /home/marius/.julia/v0.6/Atom/src/eval.jl:50
withpath at /home/marius/.julia/v0.6/CodeTools/src/utils.jl:30
unknown function (ip: 0x7fe04d48f6a2)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
withpath at /home/marius/.julia/v0.6/Atom/src/eval.jl:38
unknown function (ip: 0x7fe04d48f386)
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
macro expansion at /home/marius/.julia/v0.6/Atom/src/eval.jl:49 [inlined]
#48 at ./task.jl:80
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
jl_apply at /home/marius/src/julia/src/julia.h:1422 [inlined]
start_task at /home/marius/src/julia/src/task.c:267
unknown function (ip: 0xffffffffffffffff)

macro expansion at /home/marius/.julia/v0.6/Atom/src/eval.jl:49 [inlined]
#48 at ./task.jl:80
jl_call_fptr_internal at /home/marius/src/julia/src/julia_internal.h:337 [inlined]
jl_call_method_internal at /home/marius/src/julia/src/julia_internal.h:356 [inlined]
jl_apply_generic at /home/marius/src/julia/src/gf.c:1930
jl_apply at /home/marius/src/julia/src/julia.h:1422 [inlined]
start_task at /home/marius/src/julia/src/task.c:267
unknown function (ip: 0xffffffffffffffff)
```

```
julia> versioninfo()
Julia Version 0.6.0-rc1.0
Commit 6bdb395* (2017-05-07 00:00 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, skylake)
```

I see a possible bug in that assertion, which might be avoided by the following patch. However, I don't know that this is what you are seeing.
```julia
diff --git a/base/inference.jl b/base/inference.jl
index 9dab296025..03a7f2e5f0 100644
--- a/base/inference.jl
+++ b/base/inference.jl
@@ -704,8 +704,14 @@ function limit_type_depth(t::ANY, d::Int, cov::Bool, vars::Vector{TypeVar}=TypeV
     P = t.parameters
     isempty(P) && return t
     if d > MAX_TYPE_DEPTH
-        cov && return t.name.wrapper
-        var = TypeVar(:_, t.name.wrapper)
+        widert = t.name.wrapper
+        if !(t <: widert)
+            # oops, this'll trigger our subtype assertion
+            # widen even faster to something safe
+            widert = Any
+        end
+        cov && return widert
+        var = TypeVar(:_, widert)
         push!(vars, var)
         return var
     end
```
If not, a docker example sounds like a good idea and you can PM me a link.

Test case for triggering above error:
```julia
julia> Core.Inference.limit_type_depth(Ref{Complex{T} where T}, Core.Inference.MAX_TYPE_DEPTH)
ERROR: AssertionError: !(t isa Type) || t <: r
```

--- a/base/inference.jl
+++ b/base/inference.jl
@@ -707,8 +707,19 @@ function limit_type_depth(t::ANY, d::Int, cov::Bool, vars::Vector{TypeVar}=TypeV
     P = t.parameters
     isempty(P) && return t
     if d > MAX_TYPE_DEPTH
-        cov && return t.name.wrapper
-        var = TypeVar(:_, t.name.wrapper)
+        if isvarargtype(t)
+            # never replace Vararg with non-Vararg
+            return Vararg{limit_type_depth(P[1], d, cov, vars), P[2]}
+        end
+        widert = t.name.wrapper
+        if !(t <: widert)
+            # This can happen when a typevar has bounds too wide for its context, e.g.
+            # `Complex{T} where T` is not a subtype of `Complex`. In that case widen even
+            # faster to something safe to ensure the result is a supertype of the input.
+            widert = Any
+        end
+        cov && return widert
+        var = TypeVar(:_, widert)
         push!(vars, var)
         return var
     end

--- a/test/inference.jl
+++ b/test/inference.jl
@@ -777,3 +777,14 @@ function break_21369()
     end
 end
 @test_throws ErrorException break_21369()  # not TypeError
+
+# issue #21848
+@test Core.Inference.limit_type_depth(Ref{Complex{T} where T}, Core.Inference.MAX_TYPE_DEPTH) == Ref
+let T = Tuple{Tuple{Int64, Void},
+              Tuple{Tuple{Int64, Void},
+                    Tuple{Int64, Tuple{Tuple{Int64, Void},
+                                       Tuple{Tuple{Int64, Void}, Tuple{Int64, Tuple{Tuple{Int64, Void}, Tuple{Tuple, Tuple}}}}}}}}
+    @test Core.Inference.limit_type_depth(T, 0) >: T
+    @test Core.Inference.limit_type_depth(T, 1) >: T
+    @test Core.Inference.limit_type_depth(T, 2) >: T
+end

